### DISCUSSION TITLE
Bible browser loading time of large books

### DISCUSSION THREAD

**Discussion by zhuiks** [ezra-bible-app/ezra-bible-app#248] [General]
I can't wait 3-4s! It's too much! Is it loading the whole book or something?... :smile: 

It would be nice to have "lazy loading" for a large Bible book implemented. Lag especially feels when you just switch between the tabs

**Comment by tobias-klein** [comment]
Note that technically there's a big difference between the initial loading and the returning to an existing tab. When returning to a tab the thing that takes so long is not any additional loading process, but simply the redrawing of the existing DOM nodes of that tab. Running display: block on a large set of DOM nodes is rather slow. Of course the tab switching time would be better if there wouldn't be so many DOM nodes in the first place. Anyway ... I agree to the general request of this issue. But the solution may not be so easy. I can imagine that lazy loading would also introduce the potential of many new and strange bugs ... At least on Electron an alternative could be the use of separate web views per tab (each web view has its own DOM and hiding/showing would be much faster), but again that would introduce new potential for various bugs due to the additional complexity. What I like about the current solution is that it's rather simple from a programming perspective. And I like to keep things as simple as somehow possible.

**Comment by zhuiks** [comment]
How fast is `ipcNsi.getBookText`? 
What if we initially request only a certain amount of verses and request more verses upon scrolling?
Of course, there would be an additional challenge with chapter navigation and we either lose the scrollbar or have to do something custom

**Comment by tobias-klein** [comment]
As long as the exact technical solution is not so clear, I am turning this into a discussion ...

**Comment by tobias-klein** [comment]
The Bible text is loaded via ipcNsi / node-sword-interface in two steps. Initially the first 50 verses are loaded to quickly fill the screen (impacts visual loading performance). Then right after that the rest of the book is loaded.

Here are the numbers for the pure ipcNsi / node-sword-interface performance on two books (Philemon ... very small ... vs Psalms ... very big).

Measured on a Mac mini 2014 with the NASB module ... (not the fastest computer, but faster than an Android tablet!).

## Philemon

Part 1: 35ms
Part 2: 429ms

## Psalms

Part 1: 88ms
Part 2: 3273ms

Note that on top of this we need to consider the rendering time ...

**Reply by zhuiks** [reply]
Thanks for the timed data

I saw two parts. I guess everything that less than 100ms could be considered unnoticeable for the user.
I have to think through my idea more thoroughly, and then I'll post it here.
